docs: Add JupyterHub (self-managed) #669 (#682)
elastisys/compliantkubernetes: docs/user-guide/self-managed-services/jupyterhub.md

JupyterHub (self-managed)
===========

{%
   include-markdown './_common.include'
   start='<!--disclaimer-start-->'
   end='<!--disclaimer-end-->'
%}

> JupyterHub brings Jupyter Notebooks to the cloud. It gives the users access to computational environments and resources without burdening users with installation and maintenance tasks. This documents shows a guide on how to setup JupyterHub in a Compliant Kubernetes cluster.

Reviewer: ```suggestion
JupyterHub brings Jupyter Notebooks to the cloud. It gives the users access to computational environments and resources without burdening users with installation and maintenance tasks. This documents shows a guide on how to setup JupyterHub in a Compliant Kubernetes environment.
```

And besides that nitpick: what is required of the environment? GPUs? Auto-scaling? :thinking:

Reviewer 2: Nothing!!! :smile: This works right now on long-running.

Of course, you save tons of resources if you have auto-scaling, and you can run GPU notebooks if you have GPUs.

We'll deal with these in a second increment.

            registry.k8s.io/pause:$PAUSE_TAG traefik:$TRAEFIK_TAG
do
docker pull $IMAGE
docker tag $IMAGE  $REGISTRY/$REGISTRY_PROJECT/${IMAGE#*/}
docker push $REGISTRY/$REGISTRY_PROJECT/${IMAGE#*/}
done
```

## Configure and Deploy JupyterHub
We chose to work with [official Helm charts](https://hub.jupyter.org/helm-chart/) provided by JupyterHub. They can be downloaded via the webpage or added to your local repository by running:
```sh
helm repo add jupyterhub https://hub.jupyter.org/helm-chart/
helm repo update
```

### Configuring JupyterHub

Below is a sample **values.yaml** file that can be used to deploy JupyterHub, please read the notes and change what is necessary. This sample uses [Google OAuth](https://z2jh.jupyter.org/en/stable/administrator/authentication.html#google) for authentication and authorization.

**NOTE** Requested resources should be evaluated and reconsidered for production. Information about resource allocation and also the enabling of GPU usage can be found [here](https://z2jh.jupyter.org/en/stable/jupyterhub/customizing/user-resources.html#customizing-user-resources)

> ```yaml

Reviewer 2: I needed to add:

```
imagePullSecrets:
  - pull-secret
```

This is because my Harbor project is not public and `hook-image-awaiter-bh7fv` uses a non-default service account, which overrides the default pull secret.

Reviewer 2: I still need this. Hub seems to run under its own `hub` ServiceAccount, hence it's not using the pull secret of the `default` ServiceAccount.

hub:
  revisionHistoryLimit:
  config:
    GoogleOAuthenticator:
        client_id: $YOUR_CLIENT_ID # replace this

>         client_secret: $YOUR_CLIENT_SECRET # replace this

Reviewer 2: Please explain that we assessed that the so-called `client_secret` is not so secret: https://elastisys.io/compliantkubernetes/operator-manual/credentials/#dex-openid-client-secret

Otherwise, our application developers will freak out for leaving it in the values file.

Reviewer: Let's dream here.

What would it take for us to let our customers use the Dex from our platform to handle logins?

A well-known redirect URL, which we would make known to Dex, and that's where JupyterHub has to be deployed? And then the id/secret pair which isn't so secret?

Reviewer: I see your logic. I don't agree with you, but I see the logic behind what you're saying.

In e.g. SealedSecrets, we already make certain "holes" with names that we pre-fill for users who want to self-manage a service. There, we create a ClusterRole, for instance, and a ServiceAccount, so that they can easily use them.

Dictating here that "if you want your JupyterHub to use the same single-sign-on system as the rest of the platform, make sure you set your domain name to `jupyterhub.$DOMAIN`" doesn't sound very strange at all. And like it would add just a single line of config to our Dex.

I'm not going to rise up and fight over this, because as I said: I see your reasoning.

But as someone who cares about Developer Experience,  I don't like users having to configure their own OpenID Connect thingy if we literally have one, right there, ready to go with all kinds of config already in place to work with their identity provider of choice.

        oauth_callback_url: https://$PROJECT_DOMAIN/hub/oauth_callback # replace this
        hosted_domain:

>           - $PROJECT_DOMAIN # replace this

Reviewer: These settings are always such a pain in the behind. Please help the reader understand with what they need to replace these placeholders. 

Is $PROJECT_DOMAIN the same as $PROJECT_HOST below, on line 101?

Author: Adding additional instructions for the google oauth settings.

$PROJECT_DOMAIN and $PROJECT_HOST should be the same, changing all to $PROJECT_DOMAIN

        login_service: Google
        allow_all: true
    JupyterHub:
      admin_access: true
      authenticator_class: google
      admin_users:

>         - email@admin # replace this

Reviewer 2: ```suggestion
        - email@example.com # replace this
```

Let's stick to [RFC 2606](https://www.rfc-editor.org/rfc/rfc2606.html).

  image:
    name: $REGISTRY/$REGISTRY_PROJECT/k8s-hub:$TAG # replace this
  resources: &resourceDefaults # (1)
    requests:
      memory: 512Mi
      cpu: 10m
    limits:
      memory: 1Gi
      cpu: 1
  containerSecurityContext: &SCDefaults # (2)
    capabilities:
      drop: ["ALL"]
    runAsNonRoot: true
    seccompProfile:
      type: "RuntimeDefault"

proxy:
  service:
    type: ClusterIP
  chp:
    containerSecurityContext: *SCDefaults
    image:

>       name: $REGISTRY/$REGISTRY_PROJECT/configurable-http-proxy:$TAG  # replace this

Reviewer 2: This has several issues:

1. It seems like the Chart takes the container image tag via a separate `image.tag` value.
2. It's really painful for the user to put the right tag in the right place. Can we avoid having to specify the right tags to begin with? I see two options: (a) We extract the image tags from the `helm template`, say using grep. (b) We pin the Chart version in this doc. I'd go for (a).

Please apply to all lines with `$TAG`.

    resources: *resourceDefaults
  pause:
    image:
      name: $REGISTRY/$REGISTRY_PROJECT/pause:$TAG # replace this
    containerSecurityContext: *SCDefaults

ingress:
  enabled: true
  annotations:
    cert-manager.io/cluster-issuer: letsencrypt-prod
  ingressClassName: "nginx"
  hosts:
    - $PROJECT_DOMAIN # replace this
  tls:
    - hosts:
        - $PROJECT_DOMAIN # replace this
      secretName: jupyter-secret
```

1. The following resources are reused using *resourceDefaults later in this file
2. The following containerSecurityContext is reused using *SCDefaults later in this file
3.  Block set to true will append a privileged initContainer using the iptables to block the sensitive metadata server at the provided ip. Privileged containers are not allowed in ck8s.
4. "type: none" disables persistent storage for the user labs. Consolidate with platform administrator before enabling this feature. [for reference](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/1ebca266bed3e2f38332c5a9a3202f627cba3af0/jupyterhub/values.yaml#L383)

### Deploying JupyterHub

To deploy simply use this command in combination with the modified values.yaml as provided above.
```sh

> helm upgrade --install jupyterhub jupyterhub/jupyterhub --values values.yml

Reviewer 2: ```suggestion
helm upgrade --install jupyterhub jupyterhub/jupyterhub --values values.yaml
```

Since we use `values.yaml` above.

```

## Known Issues / Limitations

> - Scheduling is disabled because of the worker node restrictions currently in CK8s

Reviewer: ```suggestion
- Scheduling is disabled because of the worker node restrictions currently in Compliant Kubernetes
```

Did the hook not complain about this?

Reviewer: Also, help the reader understand what this means. What is the scheduling they are missing out on, and how does it affect them?

Author: expanding on this

> - JupyterHub Admin functionality is limited since it requires container root access.

Reviewer: Limited in which way?

> - Installing hub-wide python packages is not possible since requires container root access.

Reviewer: What does this mean, in practice? Can't you install as a user with pip? Or will JupyterHub just try as root, site-wide, and fail? 

Can one work around this by adding those packages in a custom image? Which, if so?

> - Persistent storage is currently disabled but [can be enabled](https://z2jh.jupyter.org/en/stable/jupyterhub/customizing/user-storage.html) depending on your infrastructure provider.

Reviewer: Please enable it. All our infra providers have a storage solution.

Reviewer 2: I specifically requested for it to be disabled. Jupyter will create one PVC per user. We don't want this to be enabled without the Jupyter admin knowing what they are doing. But yes, let's add more details on why we recommend it to be disabled as a starting point.

Reviewer: OK, let's just add more detail! :)

Author: expanding on this

> - GPU is not enabled currently but [can be enabled](https://z2jh.jupyter.org/en/stable/jupyterhub/customizing/user-resources.html#set-user-gpu-guarantees-limits) depending on your infrastructure provider.

Reviewer: This is not entirely true, yet, though, is it? I mean, we don't support them just flipping a documented switch.

@cristiklein, can we say/promise something here w.r.t. where GPU and autoscaling support is on our roadmap?

Reviewer 2: This line refers specifically to enabling GPU support in JupyterHub. It would require configuring something that Jupyter calls a new profile.

Yes, the plan is to enable this in the next increment.

![Keycloak Image](img/jupyter.gif)

## Configure and Deploy JupyterHub
We chose to work with [official Helm charts](https://hub.jupyter.org/helm-chart/) provided by JupyterHub. They can be downloaded via the webpage or added to your local repository by running:
```sh
helm repo add jupyterhub https://hub.jupyter.org/helm-chart/
helm repo update
```

### Configuring JupyterHub

Below is a sample **values.yaml** file that can be used to deploy JupyterHub, please read the notes and change what is necessary. This sample uses [Google OAuth](https://z2jh.jupyter.org/en/stable/administrator/authentication.html#google) for authentication and authorization.

**NOTE** Requested resources should be evaluated and reconsidered for production. Information about resource allocation and also the enabling of GPU usage can be found [here](https://z2jh.jupyter.org/en/stable/jupyterhub/customizing/user-resources.html#customizing-user-resources)
```yaml
hub:
  revisionHistoryLimit:
  config:
    GoogleOAuthenticator: #(5)
        client_id: $YOUR_CLIENT_ID # replace this

>         client_secret: $YOUR_CLIENT_SECRET # (6) replace this

Reviewer 2: Can you please add a comment that we determined that the OpenID client secret is not so secret? See [this](https://elastisys.io/compliantkubernetes/operator-manual/credentials/#dex-openid-client-secret).

Reviewer 3: We opted to use the plus sign annotation to place the comment about the secrets, similar to the keycloak entry. It can be found at line 145. Is this sufficient or should it be more visible?

Author: ![image](https://github.com/elastisys/compliantkubernetes/assets/143416463/43c8cfd3-2023-46b3-a42e-f2e520cfa153)

Reviewer 2: Excellent, thanks!

        oauth_callback_url: https://$PROJECT_DOMAIN/hub/oauth_callback # replace this
        hosted_domain:

>           - $PROJECT_DOMAIN # replace this

Reviewer 2: This doesn't seem to be the $PROJECT_DOMAIN. Rather it's the list of email domains allowed.

```suggestion
          - example.com # replace this the email domain which are allowed to log in to JupyterHub
```

4. "type: none" disables persistent storage for the user labs. Consolidate with platform administrator before enabling this feature. [for reference](https://github.com/jupyterhub/zero-to-jupyterhub-k8s/blob/1ebca266bed3e2f38332c5a9a3202f627cba3af0/jupyterhub/values.yaml#L383)
5. Use [this guide](https://z2jh.jupyter.org/en/stable/administrator/authentication.html#google) to get your client_id and client_secret through the [Google API Console](https://console.developers.google.com/).
6. This should not be treated as a secret. See risk analysis [here](https://github.com/dexidp/dex/issues/469) and [here](https://security.stackexchange.com/questions/225809/what-is-the-worst-i-can-do-if-i-know-openid-connect-client-secret).

### Pushing the JupyeterHub Images to Harbor
This sections shows how to pull the required images for JupyterHub and push them to another registry. If you are using the managed Harbor as your container registry, please follow [these instructions](../deploy.md) on how to authenticate, create a new project, and how to create a robot account and using it in a pull-secret to be able to pull an image from Harbor to your cluster safely.

> **NOTE** Run the following commands in the same directory as the location of your values.yaml file, since it will automatically update it with the correct images. If not, images will need to be manually set in the values.yaml.

Reviewer 2: ```suggestion
!!! note
    Run the following commands in the same directory as the location of your values.yaml file, since it will automatically update it with the correct images. If not, images will need to be manually set in the values.yaml.
```

```sh
DOMAIN=long-running.dev-ck8s.com # Replace this
REGISTRY=harbor.$DOMAIN
REGISTRY_PROJECT=jupyterhub

IMAGES=$(helm show chart jupyterhub/jupyterhub | grep image: | awk  '{print $3}')

for IMAGE in $IMAGES
do

docker pull $IMAGE
docker tag $IMAGE  $REGISTRY/$REGISTRY_PROJECT/${IMAGE#*/}
docker push $REGISTRY/$REGISTRY_PROJECT/${IMAGE#*/}
IMAGE_NAME=${IMAGE#*/} # Remove repository information
sed -i 's|name: '"${IMAGE_NAME%:*}"'|name: '"$REGISTRY/$REGISTRY_PROJECT/${IMAGE_NAME%:*}"'|g' values.yaml

done
```

### Deploying JupyterHub

> To deploy simply use this command in combination with the modified values.yaml as provided above.

Reviewer 2: ```suggestion
To deploy simply use this command in combination with the modified `values.yaml` as provided above.
```

```sh
helm upgrade --install jupyterhub jupyterhub/jupyterhub --values values.yaml
```

## Known Issues / Limitations

> - Jupyterhubs custom user scheduler is disabled (which [can help with efficient node downscaling](https://z2jh.jupyter.org/en/latest/administrator/optimization.html#scaling-down-efficiently)). For safety reasons, developers in Compliant Kubernetes do not have the rights required to deploy it.

Reviewer 2: ```suggestion
- JupyterHub's custom user scheduler is disabled (which [can help with efficient node downscaling](https://z2jh.jupyter.org/en/latest/administrator/optimization.html#scaling-down-efficiently)). For safety reasons, developers in Compliant Kubernetes do not have the rights required to deploy it.
```
